lib/chromic_pdf/assigns_plug.ex

```diff
@@ -0,0 +1,145 @@
+if Code.ensure_loaded?(Plug) and Code.ensure_loaded?(Plug.Crypto) do
+  defmodule ChromicPDF.AssignsPlug do
+    @moduledoc """
+    This module implements an "assigns passing" mechanism between a `print_to_pdf/2` caller process
+    and an internal endpoint.
+
+    ## Usage
+
+    On the caller side:
+
+        ChromicPDF.print_to_pdf({:url, "http://localhost:4000/makepdf"}, assigns: %{hello: :world})
+
+    In your endpoint:
+
+        plug ChromicPDF.AssignsPlug
+
+    The plug makes the assigns available in `conn.assigns` to be used in the HTML template.
+    """
+
+    @behaviour Plug
+
+    import Plug.Conn, only: [assign: 3]
+    alias Plug.{Conn, Crypto}
+
+    @type url :: binary
+    @type assigns :: map
+
+    @default_max_age 3600
+
+    # Default secret_key_base is generated at compile-time which ties the running Chrome instance
+    # to the compiled module.
+    # Potentially this loses requests at the edges when using _external_ chrome instances (i.e.
+    # accessed via TCP) in a clustered environment. Unlikely to be a problem, I think.
+    @default_secret_key_base 32
+                             |> :crypto.strong_rand_bytes()
+                             |> Base.encode64(padding: false)
+                             |> binary_part(0, 32)
+
+    @doc false
+    @spec secret_key_base :: binary
+    def secret_key_base do
+      Application.get_env(:chromic_pdf, :secret_key_base, @default_secret_key_base)
+    end
+
+    @doc false
+    @spec pass_assigns_via_url(url, assigns) :: url
+    def pass_assigns_via_url(url, assigns) do
+      assigns_from =
+        assigns
+        |> start_server_and_sign_pid()
+        |> encode_assigns_from()
+
+      append_query(url, %{assigns_from: assigns_from})
+    end
+
+    defp start_server_and_sign_pid(assigns) do
+      # Salt is used in the signature as well as to authorize the incoming request with the server.
+      salt = :crypto.hash(:sha256, :crypto.strong_rand_bytes(32))
+
+      {:ok, pid} = Agent.start_link(fn -> {salt, assigns} end)
```

Comment:
So `start_link` means here, we attach to the process that called the `print_to_pdf`, right? 

At the moment we are calling the `Agent.stop` when we get back the assigns in the plug. 
But that also means, that when the server that should render the url is down, we leave the Agent behind, right? 
Wouldn't it be better suited to call this in some step after the pdf generation attempt, or make the agent time out (after all the pid will only be valid for 3600).

Reply:
> But that also means, that when the server that should render the url is down, we leave the Agent behind, right?

As the process is linked, it'll die alongside the caller. Which in the scenario you describe will likely crash due to the exception from `print_to_pdf`, so I don't think that's much of a problem.

> At the moment we are calling the Agent.stop when we get back the assigns in the plug.

:point_up: Considering the point above, we could also ditch this as well. Just let it be terminated with the caller no matter what.

Matter of fact, I think I've added the `Agent.stop` call as some sorts of countermeasure in case somebody tries to bruteforce their way into the agent. But perhaps that's nonsense?

> Wouldn't it be better suited to call this in some step after the pdf generation attempt, or make the agent time out (after all the pid will only be valid for 3600).

Don't understand

Reply:
hmm, maybe keep `Agent.stop` to not accumulate these in case the client process is for some reason reused and calls `print_to_pdf` repeatedly? Seems unlikely, but the `Agent.stop` line doesn't hurt IMO